src/python/zquantum/core/circuit/conversions/pyquil_conversions.py

```diff
@@ -203,3 +221,51 @@ def convert_circuit_to_pyquil(
         program += convert_to_pyquil(gate, program)
 
     return program
+
+
+@singledispatch
+def convert_from_pyquil(obj: Union[pyquil.Program, pyquil.quil.Gate]):
+    pass
+
+
+@convert_from_pyquil.register
+def convert_gate_from_pyquil(gate: pyquil.quil.Gate) -> Gate:
+    number_of_control_modifiers = gate.modifiers.count("CONTROLLED")
+
+    all_qubits = pyquil_qubits_to_numbers(gate.qubits)
+    control_qubits = all_qubits[:number_of_control_modifiers]
+    original_qubits = all_qubits[number_of_control_modifiers:]
+
+    try:
+        gate_cls = PYQUIL_NAME_TO_ORQUESTRA_CLS[gate.name]
+        if (
+            gate_cls in SINGLE_QUBIT_NONPARAMETRIC_GATES
+            or gate_cls in TWO_QUBIT_CONTROLLED_GATES
+        ):
+            result = gate_cls(*original_qubits)
+        elif gate_cls in ROTATION_GATES or gate_cls == CPHASE:
+            result = gate_cls(
+                *original_qubits,
+                translate_expression(
+                    expression_from_pyquil(gate.params[0]), SYMPY_DIALECT
+                ),
+            )
+        elif gate_cls == SWAP:
+            result = gate_cls(original_qubits)
```

Comment:
This may be something I'm just unaware of, but why are you passing `*original_qubits` in the other cases, but `original_qubits` here?

Reply:
This is because SWAP gate has different initializer than other two-qubit gates. I think the signatures were already present in ddc4ce35ac57473eb4c811ce33105d46a27d7024 and I didn't recognize a need for uniformizing the signatures until this very point. This difference in signatures also prevents us from implementing this better logic that you suggested above. What do you think about changing the signature of SWAP gate's init so that it matches other two-qubit gates?